### PR TITLE
Gamma decay models and polarization actor update

### PR DESCRIPTION
**Pull request by BlurredChoise**

### CODE REVIEW

File: source/physics/src/GateVSource.cc [hunk 1]
```diff
@@ -746,7 +648,15 @@ void GateVSource::GeneratePrimaryVertex( G4Event* aEvent )
     {
       G4ThreeVector particle_position;
       if(mIsUserFluenceActive) { particle_position = UserFluencePosGenerateOne(); }
-      else { particle_position = m_posSPS->GenerateOne(); }
+      else {
+
+
+    	  particle_position = m_posSPS->GenerateOne();
+    	  //G4cout <<"::::CALL : m_posSPS->GenerateOne()\n";
```
Comment (wkrzemien) [inline source/physics/src/GateVSource.cc]: delete commented code

File: source/digits_hits/include/GateJPETActor.hh [hunk 2]
```diff
@@ -185,6 +208,9 @@ protected:
 	TFile * pFile;
```
Comment (wkrzemien) [inline source/digits_hits/include/GateJPETActor.hh]: Probably it was in the old code, but are those field pFile and pListeVar, pMessenger initialized in the constructor?

File: source/digits_hits/include/GateJPETActorMessenger.hh [hunk 3]
```diff
@@ -41,6 +56,7 @@ protected:
 	G4UIcmdWithABool* pEnableProcessNameCmd;
 	G4UIcmdWithABool* pEnableEmissionPointCmd;
 	G4UIcmdWithABool* pEnablePrimaryEnergyCmd;
+	G4UIcmdWithAString* pEnableASCIFileNameCmd;
```
Comment (wkrzemien) [inline source/digits_hits/include/GateJPETActorMessenger.hh:59]: Is is initialized in the constructor?

File: source/digits_hits/include/GateJPETGammaPolarizationActor.hh [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+		Limes,
+		Epsilon
+	};
+
+	virtual ~GateJPETGammaPolarizationActor();
+	FCT_FOR_AUTO_CREATOR_ACTOR(GateJPETGammaPolarizationActor)
+	// Constructs the sensor
+	virtual void Construct();
+	// Callbacks
+	virtual void UserSteppingAction(const GateVVolume *, const G4Step*);
```
Comment (wkrzemien) [inline source/digits_hits/include/GateJPETGammaPolarizationActor.hh]: not necessary but if could also use override keyword for the methods that are overridden

File: source/digits_hits/include/GateJPETGammaPolarizationActor.hh [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+	virtual void PreUserTrackingAction(const GateVVolume *, const G4Track*);
+	virtual void BeginOfEventAction(const G4Event * e);
+
+	/** Saves the data collected to the file
+	 */
+	virtual void SaveData();
```
Comment (wkrzemien) [inline source/digits_hits/include/GateJPETGammaPolarizationActor.hh]: virtual void SaveData() override;

File: source/digits_hits/include/GateJPETGammaPolarizationActor.hh [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+	 * @param: particles_number - if number of scattered per event is equal particles_number then data will be saved
+	 * */
+	void SetDesiredNumberOfParticlesScatteredPerEvent(G4int particles_number);
+
+private: //Functions
+	/** Function generate perpendicular vector to vector 'a'
+	 * @param: a - vector for which we create perpendicular one
+	 * @return: perpendicular vector to 'a'
+	*/
+	G4ThreeVector SetPerpendicularVector(G4ThreeVector& a);
```
Comment (wkrzemien) [inline source/digits_hits/include/GateJPETGammaPolarizationActor.hh]: so the argument should be const G4ThreeVector& a

File: source/digits_hits/include/GateJPETGammaPolarizationActor.hh [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+
+	/* Special function just for this class.
+	* Extract values setted by Set functions.
+	* */
+	void StandardExtractFunction(const G4Step *step);
+
+	/** Function generate dsigma/dOmega as function of theta.
```
Comment (wkrzemien) [inline source/digits_hits/include/GateJPETGammaPolarizationActor.hh]: dsigma/dOmega of what distribution?

File: source/digits_hits/include/GateJPETGammaPolarizationActor.hh [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+	 * For this this function generate histogram which is histogram dsigma/dOgema divided by sin(theta) histogram.
+	 * */
+	void GenerateRealCrossSectionOfTheta();
+
+	/** Function normalize major of histograms
```
Comment (wkrzemien) [inline source/digits_hits/include/GateJPETGammaPolarizationActor.hh]: probably you should give a list of histos here? Or maybe explicitely pass a list of histos to normalize

File: source/digits_hits/include/GateJPETGammaPolarizationActor.hh [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+	void SaveToFile(G4int eventID, G4int trackID, G4double phi, G4double theta, G4double polarization, G4ThreeVector interactionPoint, G4double totalEnergy, G4ThreeVector emisionPoint, G4double primeEnergy, G4double emisionPhi, G4double emisionTheta);
+
+	/** Function calculate phi and theta for scattered gamma.
+	 * @param: k0 - prime gamma momentum direction
+	 * @param: k - scattered gamma momentum direction
+	 * @param: e0 - prime gamma polarization
+	 * @param: theta - angle between k0 and k
+	 * @param: phi - angle between e0 and k projection on plane created by e0 and e0xk0
+	 * */
+	void GetThetaAndPhi(const G4ThreeVector k0, const G4ThreeVector k, const G4ThreeVector e0, G4double& theta, G4double& phi);
```
Comment (wkrzemien) [inline source/digits_hits/include/GateJPETGammaPolarizationActor.hh]: Why here you dont pass by const reference?

File: source/digits_hits/include/GateJPETGammaPolarizationActor.hh [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+
+	/** Function calculate angle of linear polarization
+	 * @param: k0 - prime gamma momentum direction
+	 * @param: e0 - prime gamma polarization
+	 * @param: e - scattered gamma polarization
+	 * @return: linear polarization angle (in degree)
+	 * */
+	G4double GetPolaizationAngle(const G4ThreeVector k, const G4ThreeVector e0, const G4ThreeVector e);
```
Comment (wkrzemien) [inline source/digits_hits/include/GateJPETGammaPolarizationActor.hh]: the same as above

File: source/digits_hits/include/GateJPETGammaPolarizationActor.hh [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+
+	/** Function calculate difference between real and reconstructed prime polarization vectors
+	 * @param: k0 - prime gamma momentum direction
+	 * @param: k - scattered gamma momentum direction
+	 * @param: e0 - prime gamma polarization
+	 * */
+	G4double GetAngleDifference(const G4ThreeVector k, const G4ThreeVector k0, const G4ThreeVector e0);
```
Comment (wkrzemien) [inline source/digits_hits/include/GateJPETGammaPolarizationActor.hh]: the same as above

File: source/physics/src/GateVSourceMessenger.cc [hunk 12]
```diff
@@ -253,20 +247,18 @@ void GateVSourceMessenger::SetNewValue(G4UIcommand* command,G4String newValue)
       m_source->GetEneDist()->SetMinEnergy(setMinEnergycmd->GetNewDoubleValue(newValue));
   } else if( command == VisualizeCmd ) {
     m_source->Visualize(newValue);
-  }
-  else if( command == setEnergyRangecmd ) {
+  } else if( command == setEnergyRangecmd ) {
       m_source->GetEneDist()->SetEnergyRange(setEnergyRangecmd->GetNewDoubleValue(newValue));
+  } else if( command == LinearPolarizationCmd) {
+	  G4double vl;
+	  char unts[30];
```
Comment (wkrzemien) [inline source/physics/src/GateVSourceMessenger.cc]: is this guaranteed to have 30 elementes

File: source/digits_hits/include/GateJPETGammaPolarizationActor.hh (hunk 9 above)
Comment (BlurredChoise) [reply source/digits_hits/include/GateJPETGammaPolarizationActor.hh]: Becouse method **GetThetaAndPhi** modifies variables **theta** and **phi** - you can't modifie const reference objects in function.

File: source/physics/src/GateVSourceMessenger.cc (hunk 12 above)
Comment (BlurredChoise) [reply source/physics/src/GateVSourceMessenger.cc]: It's relict code just for test - here I can use function which dont care abount units.

Comment (wkrzemien) [reply source/physics/src/GateVSourceMessenger.cc]: ok, would be better

File: source/digits_hits/include/GateJPETGammaPolarizationActor.hh (hunk 8 above)
Comment (BlurredChoise) [reply source/digits_hits/include/GateJPETGammaPolarizationActor.hh]: But why? Not all histos are required to be normalised plus creating list force me to change logic of code - generaly speking I can use list of uniq_ptr's and in this case remember about **delete** is unnecessery.

File: source/digits_hits/include/GateJPETGammaPolarizationActor.hh (hunk 6 above)
Comment (BlurredChoise) [reply source/digits_hits/include/GateJPETGammaPolarizationActor.hh]: yup

File: source/digits_hits/include/GateJPETActorMessenger.hh (hunk 3 above)
Comment (BlurredChoise) [reply source/digits_hits/include/GateJPETActorMessenger.hh]: Nope - becouse whole pointers are initialized in *BuildCommands* - all messanger classes are created in this way in Gate

File: source/digits_hits/include/GateJPETGammaPolarizationActor.hh (hunk 8 above)
Comment (wkrzemien) [reply source/digits_hits/include/GateJPETGammaPolarizationActor.hh]: Ok you can leave it as it is.
In general I prefer to know what the function does. From the description you have no idea which histograms will be normalize and which not.

**Review by wkrzemien — CHANGES_REQUESTED** [review]
Please update those small things that we agreed and I will accept it

**Comment by BlurredChoise** [comment]
Done.

**Comment by wkrzemien** [comment]
@BlurredChoise: There are two conflicts:
source/physics/include/Gate3GammaAnnihilation.hh
source/physics/src/Gate3GammaAnnihilation.cc
Please remove it, cause I cannot merge your PR

**Comment by BlurredChoise** [comment]
Resolved

**Pull request merged**